apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action.ts

```diff
@@ -146,43 +165,103 @@ const updateStateFromInferredFacts = async (params: {
-      const duplicate = duplicates.find(({ duplicateIds }) =>
-        duplicateIds.includes(entityId),
-      );
+        return duplicate ? null : proposedEntity;
+      })
+      .filter(isNotNullish);
+  }
+
+  state.filesUsedToInferFacts.push(...newFilesUsedToInferFacts);
+
+  /**
+   * Step 2: Create or update proposals for new entities and existing entities with new facts
+   */
+
+  /**
+   * We want to (re)propose entities which may have new information, via one of:
+   * 1. Appearing as a new summary
+   * 2. Being the subject or object of a new fact
+   * 3. Having been identified as the canonical version of a new duplicate (which means it may have had new facts discovered)
+   */
+  const entityIdsToPropose = [
+    ...new Set([
+      ...newEntitySummaries.map(({ localId }) => localId),
+      ...newFacts.flatMap(({ subjectEntityLocalId, objectEntityLocalId }) =>
+        [subjectEntityLocalId, objectEntityLocalId].filter(isNotNullish),
+      ),
+      ...canonicalEntityIdsForNewDuplicates,
+    ]),
+  ];
+
+  /**
+   * Gather the facts which relate to the entities that are being proposed
+   */
+  const entitySummaries = state.entitySummaries.filter(({ localId }) =>
+    entityIdsToPropose.includes(localId),
+  );
+
+  const relevantFacts = state.inferredFacts.filter(
+    ({ subjectEntityLocalId, objectEntityLocalId }) =>
+      entityIdsToPropose.includes(subjectEntityLocalId) ||
+      /**
+       * Facts where the entity is the object may contain information which is useful in constructing it,
+       * or a link from it – the fact may be expressed in the reverse direction to that of the target entity types.
+       */
+      (objectEntityLocalId && entityIdsToPropose.includes(objectEntityLocalId)),
+  );
 
-      return duplicate ? duplicate.canonicalId : entityId;
+  /**
+   * Given the affected entities, we also need the summaries of entities they may link to.
+   */
+  const potentialLinkTargetEntitySummaries = state.entitySummaries.filter(
+    ({ localId }) =>
+      relevantFacts.some(
+        ({ objectEntityLocalId }) => localId === objectEntityLocalId,
+      ),
+  );
+
+  /**
+   * Get the updated proposals
+   */
+  const { proposedEntities: newProposedEntities } =
+    await proposeEntitiesFromFacts({
+      dereferencedEntityTypes: input.allDereferencedEntityTypesById,
+      entitySummaries,
+      existingEntitySummaries: input.existingEntitySummaries,
+      facts: relevantFacts,
+      potentialLinkTargetEntitySummaries,
     });
-  }
 
-  state.filesUsedToInferFacts.push(...filesUsedToInferFacts);
+  state.proposedEntities = [
+    /**
+     * Filter out any previous proposed entities that have been re-proposed with new facts.
+     */
+    ...state.proposedEntities.filter(
+      ({ localEntityId }) =>
+        !newProposedEntities.some(
+          (newProposedEntity) =>
+            newProposedEntity.localEntityId === localEntityId,
+        ),
+    ),
+    ...newProposedEntities,
+  ];
 };
```

Comment:
Just want to flag, unless I've missed something I don't think we are filtering out previously proposed entities, which have been provided with new `localEntityId`s after the deduplication process (may have been marked as a non-canonical duplicate of an entity summary which was obtained in the worker agent).

So we may end up with multiple proposals for the same entity, with different `localEntityId`s. We could update the deduplication logic to also prune proposals which have been deduplicated, or replace their `localEntityId` so they will be filtered out here.

Reply:
I think this is happening [here](https://github.com/hashintel/hash/blob/07812d8b86d9d3964493860b828dac259d9462f9/apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action.ts#L177), although I may be misunderstanding.

Worth also flagging that I changed the logic there to remove any proposed entities from state which had been marked as a duplicate, rather than overwriting their `localEntityId` (which would mean two different couples of the entity with the same `localEntityId` in state, possibly with different properties) – it is quite possible I've missed/misunderstood something about this though.

Reply:
> I think this is happening [here](https://github.com/hashintel/hash/blob/07812d8b86d9d3964493860b828dac259d9462f9/apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action.ts#L177), although I may be misunderstanding.

Ah yes, this does address the issue I flagged.

> I changed the logic there to remove any proposed entities from state which had been marked as a duplicate, rather than overwriting their localEntityId (which would mean two different couples of the entity with the same localEntityId in state, possibly with different properties)

Okay, that works but overwriting the `localEntityId` should also work, as we then deduplicate those previously proposed entities based on their `localEntityId` here. The only benefit I can think of is that if there is some error encountered re-proposing the entity, we don't lose the prior proposal. But that may not be something worth accounting for.